docs/tp/04.md

```diff
@@ -0,0 +1,60 @@
+## Resources
+ 1. [OpenOptions](https://doc.rust-lang.org/std/fs/struct.OpenOptions.html)
+ 2. [Stdin](https://doc.rust-lang.org/std/io/fn.stdin.html)
+ 3. [Stdout](https://doc.rust-lang.org/std/io/fn.stdout.html)
+ 4. [dup](https://docs.rs/nix/0.23.1/nix/unistd/fn.dup.html)
+ 5. [dup2](https://docs.rs/nix/0.23.1/nix/unistd/fn.dup2.html)
+
+## Linux Standard Streams
+
+Sous Linux, `stdin` est le flux d'entrée standard. Cela accepte le texte comme entrée. La sortie texte de la commande au shell est transmise via le flux `stdout` (sortie standard). Les messages d'erreur de la commande sont envoyés via le flux `stderr` (erreur standard).
+
+Vous pouvez donc voir qu'il existe deux flux de sortie, `stdout` et `stderr`, et un flux d'entrée, `stdin`. Étant donné que les messages d'erreur et la sortie normale ont chacun leur propre conduit pour les acheminer vers la fenêtre du terminal, ils peuvent être traités indépendamment les uns des autres.
+
+Les flux sous Linux, comme presque tout le reste, sont traités comme s'ils étaient des **fichiers**. Vous pouvez lire du texte à partir d'un fichier et vous pouvez écrire du texte dans un fichier. Ces deux actions impliquent un flux de données. Ainsi, le concept de traitement d'un flux de données en tant que fichier n'est pas si exagéré.
+
+:::note
+Chaque fichier associé à un processus se voit attribuer un **numéro unique** pour l'identifier. C'est ce qu'on appelle le **descripteur de fichier**. Chaque fois qu'une action doit être effectuée sur un fichier, le descripteur de fichier est utilisé pour identifier le fichier.
+:::
+
+Ces valeurs sont toujours utilisées pour stdin, stdout et stderr :
+* 0: stdin
+* 1: stdout
+* 2: stderr
+
+## Les appels système dup et dup2
+
+### dup
+L'appel système `dup()` crée **une copie** d'un descripteur de fichier.
+* Il utilise le descripteur inutilisé portant le numéro le plus bas pour le nouveau descripteur.
+* Si la copie est créée avec succès, les descripteurs de fichier d'origine et de copie peuvent être utilisés de manière interchangeable.
+* Ils se réfèrent tous les deux à la même description de fichier ouvert et partagent donc les indicateurs de décalage de fichier et d'état de fichier.
+
+### dup2
+L'appel système `dup2()` est similaire à `dup()` mais la différence fondamentale entre eux est qu'au lieu d'utiliser le descripteur de fichier inutilisé portant le numéro le plus bas, il utilise le numéro de descripteur **spécifié par l'utilisateur**.
+
+:::info
+Lors de l'appel de [dup2](https://docs.rs/nix/0.23.1/nix/unistd/fn.dup2.html) dans Rust, les paramètres donnés doivent être l'ancien descripteur de fichier et le nouveau.
+:::
+
+### Redirections
+Quand vous voulez rediriger stdout vers un fichier vous pouvez le faire en ouvrant un fichier à l'aide de l'appel système `open()`, puis en dupliquant ce descripteur de fichier sur `stdout` en utilisant `dup2()`, de sorte que la sortie standard soit maintenant représentée par le fichier ouvert et non par la console.
```

Comment:
We need an example with `as_fd` instead of `open`.

Reply:
Done